pilot/ARCHITECTURE.md

```diff
@@ -0,0 +1,163 @@
+
+Primarily for historical reasons, ingestion is split into a few components.
+
+#### ConfigStore
+
+The `ConfigStore` reads a variety of resources and exposes them over a standard interface (Get, List, etc). These types are wrapped in a common `config.Config` struct, contrasting with typical Kubernetes clients which use per-resource types. The most common is reading from Kubernetes via the `crdclient` package.
+
+```mermaid
+graph TD
+	subgraph ConfigStore
+		xcs(XDS Client)
+		ccs(CRD Client)
+		fcs(Filesystem Client)
+		acs(Aggregate)
+		xcs-->acs
+		ccs-->acs
+		fcs-->acs
+	end
+```
+#### ServiceDiscovery
+
+The other primary interface is the ServiceDiscovery. Similar to ConfigStore, this aggregates over a variety of resources. However, it does not provide generic resource access, and instead precomputes a variety of service-oriented internal resources, such as `model.Service` and `model.ServiceInstance`.
+
+This is composed of two controllers - one driven from core Kubernetes types ("Kube Controller") and one by Istio types ("ServiceEntry controller").
+
+```mermaid
+graph TD
+	subgraph Kube Controller
+		s(Services)
+		e(Endpoints)
+		p(Pods)
+		ksi(ServiceInstances)
+		kwi(WorkloadInstances)
+		s-->ksi
+		e-->ksi
+		p-->kwi
+		
+	end
+	subgraph ServiceEntry Controller
+		se(ServiceEntry)
+		we(WorkloadEntry)
+		ssi(ServiceInstances)
+		swi(WorkloadInstances)
+		se-->ssi
+		swi-->ssi
+		we-->swi
+	end
+	kwi-->ssi
+	swi-->ksi
+```
+For the most part this is fairly straight forward. However, we support `ServiceEntry` selecting `Pod`, and `Service` selecting `WorkloadEntry`, which leads to cross-controller communication.
+
+Note: the asymmetry with `Pods` not contributing to Kube controller's `ServiceInstances` is due to the use of `Endpoints` which itself is derived from `Pod` from Kubernetes core.
+
+#### PushContext
+`PushContext` is an immutable snapshot of the current state of the world. It is regenerated (usually partially) on each configuration push (more on this below). Due to being a snapshot, most lookups are lock-free.
+
+`PushContext` is built up by querying the above layers. For some simple use cases, this is as simple as storing something like `configstore.List(SomeType)`; in this case, the only difference from directly exposing the configstore is to snapshot the current state. In other cases, some pre-computations and indexes are computed to make later accesses efficient.
+
+#### Endpoints
+Endpoints have an optimized code path, as they are by far the most frequently updated resource - in a steady cluster, this will often be the *only* change, caused by scale up/down.
+
+As a result, they do not go through `PushContext`, and changes do not trigger a `PushContext` recomputation. Instead, the current state is incrementally computed based on events from `ServiceDiscovery`.
```

Comment:
So roughly:

PushContext == (ServiceDiscovery + ConfigStore)

Endpoints == ServiceDiscovery?

Reply:
yep

Reply:
Also endpoints are latency critical ( you want traffic as soon as pod is ready - and to stop when not healthy) and will not need rollback or controlled distribution ( the other idea behind PushContext was to slow it down and send it first to canaries, etc).